File: android/app/src/main/java/com/nearme/app/viewmodels/NotificationViewModels.kt
package com.nearme.app.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.nearme.app.models.*
import com.nearme.app.services.NotificationService
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import java.util.*

// MARK: - Notification History View Model
class NotificationHistoryViewModel : ViewModel() {
    private val _notifications = MutableStateFlow<List<NotificationItem>>(emptyList())
    val notifications: StateFlow<List<NotificationItem>> = _notifications.asStateFlow()
    
    private val _selectedFilter = MutableStateFlow(NotificationFilter.ALL)
    val selectedFilter: StateFlow<NotificationFilter> = _selectedFilter.asStateFlow()
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    
    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()
    
    val filteredNotifications: StateFlow<List<NotificationItem>> = combine(
        _notifications,
        _selectedFilter
    ) { notifications, filter ->
        notifications.filter { notification ->
            when (filter) {
                NotificationFilter.ALL -> true
                NotificationFilter.UNREAD -> !notification.isRead
                NotificationFilter.TODAY -> isToday(notification.timestamp)
                NotificationFilter.THIS_WEEK -> isThisWeek(notification.timestamp)
                NotificationFilter.COMPLETED -> notification.type == NotificationType.COMPLETION
                NotificationFilter.SNOOZED -> notification.actions.any { it.identifier.contains("SNOOZE") }
            }
        }.sortedByDescending { it.timestamp }
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5000),
        initialValue = emptyList()
    )
    
    private val notificationService = NotificationService()
    
    init {
        loadNotifications()
    }
    
    fun loadNotifications() {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null
            
            try {
                val notifications = notificationService.getNotificationHistory()
                _notifications.value = notifications
            } catch (e: Exception) {
                _errorMessage.value = e.message
                loadSampleNotifications() // Fallback to sample data
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    fun handleAction(action: NotificationAction, notification: NotificationItem) {
        viewModelScope.launch {
            // Update local state immediately for responsive UI
            val updatedNotifications = _notifications.value.toMutableList()
            val index = updatedNotifications.indexOfFirst { it.id == notification.id }
            if (index != -1) {
                updatedNotifications[index] = updatedNotifications[index].copy(isRead = true)
                _notifications.value = updatedNotifications
            }
            
            // Handle the action
            when (action.identifier) {
                "COMPLETE_ACTION" -> handleCompleteAction(notification)
                "SNOOZE_15M_ACTION" -> handleSnoozeAction(notification, SnoozeDuration.FIFTEEN_MINUTES)
                "SNOOZE_1H_ACTION" -> handleSnoozeAction(notification, SnoozeDuration.ONE_HOUR)
                "SNOOZE_TODAY_ACTION" -> handleSnoozeAction(notification, SnoozeDuration.TODAY)
                "MUTE_ACTION" -> handleMuteAction(notification)
                "OPEN_MAP_ACTION" -> handleOpenMapAction(notification)
            }
            
            // Send action to backend
            try {
                notificationService.performNotificationAction(notification.id, action.identifier)
            } catch (e: Exception) {
                _errorMessage.value = "Failed to perform action: ${e.message}"
            }
        }
    }
    
    fun markAllAsRead() {
        viewModelScope.launch {
            val updatedNotifications = _notifications.value.map { it.copy(isRead = true) }
            _notifications.value = updatedNotifications
            
            try {
                notificationService.markAllNotificationsAsRead()
            } catch (e: Exception) {
                _errorMessage.value = "Failed to mark notifications as read: ${e.message}"
            }
        }
    }
    
    fun deleteNotification(notification: NotificationItem) {
        viewModelScope.launch {
            val updatedNotifications = _notifications.value.toMutableList()
            updatedNotifications.removeAll { it.id == notification.id }
            _notifications.value = updatedNotifications
            
            try {
                notificationService.deleteNotification(notification.id)
            } catch (e: Exception) {
                _errorMessage.value = "Failed to delete notification: ${e.message}"
                // Re-add the notification if deletion failed
                _notifications.value = _notifications.value + notification
            }
        }
    }
    
    // MARK: - Private Methods
    
    private fun handleCompleteAction(notification: NotificationItem) {
        // Remove from list
        val updatedNotifications = _notifications.value.toMutableList()
        updatedNotifications.removeAll { it.id == notification.id }
        _notifications.value = updatedNotifications
    }
    
    private fun handleSnoozeAction(notification: NotificationItem, duration: SnoozeDuration) {
        // Mark as read (already done in handleAction)
        // In a real implementation, we would schedule a snooze notification
    }
    
    private fun handleMuteAction(notification: NotificationItem) {
        // Remove from list
        val updatedNotifications = _notifications.value.toMutableList()
        updatedNotifications.removeAll { it.id == notification.id }
        _notifications.value = updatedNotifications
    }
    
    private fun handleOpenMapAction(notification: NotificationItem) {
        // In a real implementation, we would open the map
    }
    
    private fun loadSampleNotifications() {
        // Load sample notifications for development/demo purposes
        _notifications.value = listOf(
            NotificationItem(
                title = "You're near the pharmacy",
                body = "Don't forget to pick up your prescription",
                taskId = "task-1",
                type = NotificationType.APPROACH,
                actions = listOf(
                    NotificationAction.complete,
                    NotificationAction.snooze15m,
                    NotificationAction.snooze1h,
                    NotificationAction.mute
                )
            ),
            NotificationItem(
                title = "Arrived at grocery store",
                body = "Time to pick up milk and bread",
                taskId = "task-2",
                type = NotificationType.ARRIVAL,
                actions = listOf(
                    NotificationAction.complete,
                    NotificationAction.snooze15m,
                    NotificationAction.openMap
                ),
                isRead = true
            ),
            NotificationItem(
                title = "Task completed!",
                body = "Great job picking up your dry cleaning",
                taskId = "task-3",
                type = NotificationType.COMPLETION,
                isRead = true
            )
        )
    }
    
    private fun isToday(date: Date): Boolean {
        val calendar = Calendar.getInstance()
        val today = calendar.time
        calendar.time = date
        val notificationDate = calendar.time
        
        val todayCalendar = Calendar.getInstance()
        todayCalendar.time = today
        
        val notificationCalendar = Calendar.getInstance()
        notificationCalendar.time = notificationDate
        
        return todayCalendar.get(Calendar.YEAR) == notificationCalendar.get(Calendar.YEAR) &&
                todayCalendar.get(Calendar.DAY_OF_YEAR) == notificationCalendar.get(Calendar.DAY_OF_YEAR)
    }
    
    private fun isThisWeek(date: Date): Boolean {
        val calendar = Calendar.getInstance()
        val now = calendar.time
        calendar.time = date
        val notificationDate = calendar.time
        
        val nowCalendar = Calendar.getInstance()
        nowCalendar.time = now
        
        val notificationCalendar = Calendar.getInstance()
        notificationCalendar.time = notificationDate
        
        return nowCalendar.get(Calendar.YEAR) == notificationCalendar.get(Calendar.YEAR) &&
                nowCalendar.get(Calendar.WEEK_OF_YEAR) == notificationCalendar.get(Calendar.WEEK_OF_YEAR)
    }
}

// MARK: - Notification Settings View Model
class NotificationSettingsViewModel : ViewModel() {
    private val _settings = MutableStateFlow(NotificationSettings.default)
    val settings: StateFlow<NotificationSettings> = _settings.asStateFlow()
    
    private val notificationService = NotificationService()
    
    init {
        loadSettings()
    }
    
    fun loadSettings() {
        // In a real implementation, we would load from SharedPreferences or a database
        _settings.value = NotificationSettings.default
    }
    
    fun saveSettings() {
        viewModelScope.launch {
            try {
                notificationService.updateNotificationSettings(_settings.value)
            } catch (e: Exception) {
                // Handle error
            }
        }
    }
    
    fun updateLocationRemindersEnabled(enabled: Boolean) {
        _settings.value = _settings.value.copy(locationRemindersEnabled = enabled)
        saveSettings()
    }
    
    fun updateTaskCompletionAlerts(enabled: Boolean) {
        _settings.value = _settings.value.copy(taskCompletionAlerts = enabled)
        saveSettings()
    }
    
    fun updateDailySummary(enabled: Boolean) {
        _settings.value = _settings.value.copy(dailySummary = enabled)
        saveSettings()
    }
    
    fun updateQuietHoursEnabled(enabled: Boolean) {
        _settings.value = _settings.value.copy(quietHoursEnabled = enabled)
        saveSettings()
    }
    
    fun updateQuietHoursStart(time: Date) {
        _settings.value = _settings.value.copy(quietHoursStart = time)
        saveSettings()
    }
    
    fun updateQuietHoursEnd(time: Date) {
        _settings.value = _settings.value.copy(quietHoursEnd = time)
        saveSettings()
    }
    
    fun updateDefaultSnoozeDuration(duration: SnoozeDuration) {
        _settings.value = _settings.value.copy(defaultSnoozeDuration = duration)
        saveSettings()
    }
    
    fun resetToDefaults() {
        _settings.value = NotificationSettings.default
        saveSettings()
    }
}

// MARK: - Notification Service
class NotificationService {
    suspend fun getNotificationHistory(): List<NotificationItem> {
        // TODO: Implement actual API call
        // For now, return empty list
        return emptyList()
    }
    
    suspend fun performNotificationAction(notificationId: String, action: String) {
        // TODO: Implement actual API call
    }
    
    suspend fun markAllNotificationsAsRead() {
        // TODO: Implement actual API call
    }
    
    suspend fun deleteNotification(notificationId: String) {
        // TODO: Implement actual API call
    }
    
    suspend fun updateNotificationSettings(settings: NotificationSettings) {
        // TODO: Implement actual API call
    }
}

File: ios/NearMe/ViewModels/NotificationViewModels.swift
import Foundation
import Combine
import UserNotifications

// MARK: - Notification History View Model
class NotificationHistoryViewModel: ObservableObject {
    @Published var notifications: [NotificationItem] = []
    @Published var selectedFilter: NotificationFilter = .all
    @Published var isLoading = false
    @Published var errorMessage: String?
    
    private var cancellables = Set<AnyCancellable>()
    private let notificationService = NotificationService.shared
    
    var filteredNotifications: [NotificationItem] {
        let filtered = notifications.filter { notification in
            switch selectedFilter {
            case .all:
                return true
            case .unread:
                return !notification.isRead
            case .today:
                return Calendar.current.isDateInToday(notification.timestamp)
            case .thisWeek:
                return Calendar.current.isDate(notification.timestamp, equalTo: Date(), toGranularity: .weekOfYear)
            case .completed:
                return notification.type == .completion
            case .snoozed:
                return notification.actions.contains { $0.identifier.contains("SNOOZE") }
            }
        }
        
        return filtered.sorted { $0.timestamp > $1.timestamp }
    }
    
    init() {
        setupNotificationObservers()
    }
    
    private func setupNotificationObservers() {
        // Listen for new notifications
        NotificationCenter.default.publisher(for: .newNotificationReceived)
            .sink { [weak self] notification in
                self?.handleNewNotification(notification)
            }
            .store(in: &cancellables)
        
        // Listen for notification actions
        NotificationCenter.default.publisher(for: .notificationActionPerformed)
            .sink { [weak self] notification in
                self?.handleNotificationAction(notification)
            }
            .store(in: &cancellables)
    }
    
    func loadNotifications() {
        isLoading = true
        errorMessage = nil
        
        notificationService.getNotificationHistory { [weak self] result in
            DispatchQueue.main.async {
                self?.isLoading = false
                
                switch result {
                case .success(let notifications):
                    self?.notifications = notifications
                case .failure(let error):
                    self?.errorMessage = error.localizedDescription
                    self?.loadSampleNotifications() // Fallback to sample data
                }
            }
        }
    }
    
    func handleAction(_ action: NotificationAction, for notification: NotificationItem) {
        // Update local state immediately for responsive UI
        if let index = notifications.firstIndex(where: { $0.id == notification.id }) {
            notifications[index].isRead = true
        }
        
        // Handle the action
        switch action.identifier {
        case "COMPLETE_ACTION":
            handleCompleteAction(notification)
        case "SNOOZE_15M_ACTION":
            handleSnoozeAction(notification, duration: .fifteenMinutes)
        case "SNOOZE_1H_ACTION":
            handleSnoozeAction(notification, duration: .oneHour)
        case "SNOOZE_TODAY_ACTION":
            handleSnoozeAction(notification, duration: .today)
        case "MUTE_ACTION":
            handleMuteAction(notification)
        case "OPEN_MAP_ACTION":
            handleOpenMapAction(notification)
        default:
            break
        }
        
        // Send action to backend
        notificationService.performNotificationAction(
            notificationId: notification.id,
            action: action.identifier
        ) { [weak self] result in
            DispatchQueue.main.async {
                switch result {
                case .success:
                    print("Action performed successfully")
                case .failure(let error):
                    self?.errorMessage = "Failed to perform action: \(error.localizedDescription)"
                }
            }
        }
    }
    
    func markAllAsRead() {
        notifications = notifications.map { notification in
            var updatedNotification = notification
            updatedNotification.isRead = true
            return updatedNotification
        }
        
        notificationService.markAllNotificationsAsRead { [weak self] result in
            DispatchQueue.main.async {
                switch result {
                case .success:
                    print("All notifications marked as read")
                case .failure(let error):
                    self?.errorMessage = "Failed to mark notifications as read: \(error.localizedDescription)"
                }
            }
        }
    }
    
    func deleteNotification(_ notification: NotificationItem) {
        notifications.removeAll { $0.id == notification.id }
        
        notificationService.deleteNotification(notification.id) { [weak self] result in
            DispatchQueue.main.async {
                switch result {
                case .success:
                    print("Notification deleted successfully")
                case .failure(let error):
                    self?.errorMessage = "Failed to delete notification: \(error.localizedDescription)"
                    // Re-add the notification if deletion failed
                    self?.notifications.append(notification)
                }
            }
        }
    }
    
    // MARK: - Private Methods
    
    private func handleNewNotification(_ notification: Notification) {
        guard let userInfo = notification.userInfo,
              let notificationData = userInfo["notification"] as? NotificationItem else {
            return
        }
        
        notifications.insert(notificationData, at: 0)
    }
    
    private func handleNotificationAction(_ notification: Notification) {
        guard let userInfo = notification.userInfo,
              let notificationId = userInfo["notificationId"] as? String,
              let action = userInfo["action"] as? String else {
            return
        }
        
        // Update notification based on action
        if let index = notifications.firstIndex(where: { $0.id == notificationId }) {
            switch action {
            case "complete":
                notifications.remove(at: index)
            case "snooze":
                notifications[index].isRead = true
            case "mute":
                notifications.remove(at: index)
            default:
                break
            }
        }
    }
    
    private func handleCompleteAction(_ notification: NotificationItem) {
        // Post completion notification
        NotificationCenter.default.post(
            name: .taskCompleted,
            object: notification.taskId
        )
        
        // Remove from list
        notifications.removeAll { $0.id == notification.id }
    }
    
    private func handleSnoozeAction(_ notification: NotificationItem, duration: SnoozeDuration) {
        // Post snooze notification
        NotificationCenter.default.post(
            name: .taskSnoozed,
            object: [
                "taskId": notification.taskId ?? "",
                "duration": duration.rawValue
            ]
        )
        
        // Mark as read
        if let index = notifications.firstIndex(where: { $0.id == notification.id }) {
            notifications[index].isRead = true
        }
    }
    
    private func handleMuteAction(_ notification: NotificationItem) {
        // Post mute notification
        NotificationCenter.default.post(
            name: .taskMuted,
            object: notification.taskId
        )
        
        // Remove from list
        notifications.removeAll { $0.id == notification.id }
    }
    
    private func handleOpenMapAction(_ notification: NotificationItem) {
        // Post map open notification
        NotificationCenter.default.post(
            name: .openMapForTask,
            object: notification.taskId
        )
    }
    
    private func loadSampleNotifications() {
        // Load sample notifications for development/demo purposes
        notifications = [
            NotificationItem(
                title: "You're near the pharmacy",
                body: "Don't forget to pick up your prescription",
                taskId: "task-1",
                type: .approach,
                actions: [.complete, .snooze15m, .snooze1h, .mute]
            ),
            NotificationItem(
                title: "Arrived at grocery store",
                body: "Time to pick up milk and bread",
                taskId: "task-2",
                type: .arrival,
                actions: [.complete, .snooze15m, .openMap],
                isRead: true
            ),
            NotificationItem(
                title: "Task completed!",
                body: "Great job picking up your dry cleaning",
                taskId: "task-3",
                type: .completion,
                isRead: true
            )
        ]
    }
}

// MARK: - Notification Settings View Model
class NotificationSettingsViewModel: ObservableObject {
    @Published var locationRemindersEnabled: Bool = true
    @Published var taskCompletionAlerts: Bool = true
    @Published var dailySummary: Bool = false
    @Published var quietHoursEnabled: Bool = false
    @Published var quietHoursStart: Date = Calendar.current.date(bySettingHour: 22, minute: 0, second: 0, of: Date()) ?? Date()
    @Published var quietHoursEnd: Date = Calendar.current.date(bySettingHour: 8, minute: 0, second: 0, of: Date()) ?? Date()
    @Published var defaultSnoozeDuration: SnoozeDuration = .fifteenMinutes
    @Published var soundEnabled: Bool = true
    @Published var vibrationEnabled: Bool = true
    @Published var badgeEnabled: Bool = true
    
    private let settingsKey = "notification_settings"
    private let notificationService = NotificationService.shared
    
    func loadSettings() {
        if let data = UserDefaults.standard.data(forKey: settingsKey),
           let settings = try? JSONDecoder().decode(NotificationSettings.self, from: data) {
            locationRemindersEnabled = settings.locationRemindersEnabled
            taskCompletionAlerts = settings.taskCompletionAlerts
            dailySummary = settings.dailySummary
            quietHoursEnabled = settings.quietHoursEnabled
            quietHoursStart = settings.quietHoursStart
            quietHoursEnd = settings.quietHoursEnd
            defaultSnoozeDuration = settings.defaultSnoozeDuration
            soundEnabled = settings.soundEnabled
            vibrationEnabled = settings.vibrationEnabled
            badgeEnabled = settings.badgeEnabled
        }
    }
    
    func saveSettings() {
        let settings = NotificationSettings(
            locationRemindersEnabled: locationRemindersEnabled,
            taskCompletionAlerts: taskCompletionAlerts,
            dailySummary: dailySummary,
            quietHoursEnabled: quietHoursEnabled,
            quietHoursStart: quietHoursStart,
            quietHoursEnd: quietHoursEnd,
            defaultSnoozeDuration: defaultSnoozeDuration,
            soundEnabled: soundEnabled,
            vibrationEnabled: vibrationEnabled,
            badgeEnabled: badgeEnabled
        )
        
        if let data = try? JSONEncoder().encode(settings) {
            UserDefaults.standard.set(data, forKey: settingsKey)
        }
        
        // Send settings to backend
        notificationService.updateNotificationSettings(settings) { result in
            switch result {
            case .success:
                print("Notification settings updated successfully")
            case .failure(let error):
                print("Failed to update notification settings: \(error)")
            }
        }
    }
    
    func resetToDefaults() {
        locationRemindersEnabled = true
        taskCompletionAlerts = true
        dailySummary = false
        quietHoursEnabled = false
        quietHoursStart = Calendar.current.date(bySettingHour: 22, minute: 0, second: 0, of: Date()) ?? Date()
        quietHoursEnd = Calendar.current.date(bySettingHour: 8, minute: 0, second: 0, of: Date()) ?? Date()
        defaultSnoozeDuration = .fifteenMinutes
        soundEnabled = true
        vibrationEnabled = true
        badgeEnabled = true
        
        saveSettings()
    }
}

// MARK: - Notification Service
class NotificationService {
    static let shared = NotificationService()
    
    private let apiBaseURL = "http://localhost:3000/api" // TODO: Move to configuration
    private let session = URLSession.shared
    
    private init() {}
    
    func getNotificationHistory(completion: @escaping (Result<[NotificationItem], Error>) -> Void) {
        // TODO: Implement actual API call
        // For now, return sample data
        DispatchQueue.global().asyncAfter(deadline: .now() + 1) {
            completion(.success([]))
        }
    }
    
    func performNotificationAction(
        notificationId: String,
        action: String,
        completion: @escaping (Result<Void, Error>) -> Void
    ) {
        // TODO: Implement actual API call
        DispatchQueue.global().asyncAfter(deadline: .now() + 0.5) {
            completion(.success(()))
        }
    }
    
    func markAllNotificationsAsRead(completion: @escaping (Result<Void, Error>) -> Void) {
        // TODO: Implement actual API call
        DispatchQueue.global().asyncAfter(deadline: .now() + 0.5) {
            completion(.success(()))
        }
    }
    
    func deleteNotification(_ notificationId: String, completion: @escaping (Result<Void, Error>) -> Void) {
        // TODO: Implement actual API call
        DispatchQueue.global().asyncAfter(deadline: .now() + 0.5) {
            completion(.success(()))
        }
    }
    
    func updateNotificationSettings(_ settings: NotificationSettings, completion: @escaping (Result<Void, Error>) -> Void) {
        // TODO: Implement actual API call
        DispatchQueue.global().asyncAfter(deadline: .now() + 0.5) {
            completion(.success(()))
        }
    }
}

// MARK: - Notification Names
extension Notification.Name {
    static let newNotificationReceived = Notification.Name("newNotificationReceived")
    static let notificationActionPerformed = Notification.Name("notificationActionPerformed")
    static let openMapForTask = Notification.Name("openMapForTask")
}
